pinot-plugins/pinot-stream-ingestion/pinot-kinesis/src/main/java/org/apache/pinot/plugin/stream/kinesis/KinesisConsumer.java

```diff
@@ -69,134 +61,77 @@ public KinesisConsumer(KinesisConfig config, KinesisClient kinesisClient) {
+      GetShardIteratorRequest getShardIteratorRequest =
+          GetShardIteratorRequest.builder().streamName(_config.getStreamTopicName()).shardId(shardId)
+              .startingSequenceNumber(startSequenceNumber).shardIteratorType(ShardIteratorType.AFTER_SEQUENCE_NUMBER)
+              .build();
+      shardIterator = _kinesisClient.getShardIterator(getShardIteratorRequest).shardIterator();
+    }
+    if (shardIterator == null) {
+      return new KinesisMessageBatch(List.of(), startOffset, true);
     }
-  }
 
-  private KinesisMessageBatch buildKinesisMessageBatch(KinesisPartitionGroupOffset startOffset,
-      List<BytesStreamMessage> messages, boolean endOfShard) {
+    // Read records
+    rateLimitRequests();
+    GetRecordsRequest getRecordRequest =
+        GetRecordsRequest.builder().shardIterator(shardIterator).limit(_config.getNumMaxRecordsToFetch()).build();
+    GetRecordsResponse getRecordsResponse = _kinesisClient.getRecords(getRecordRequest);
+    List<Record> records = getRecordsResponse.records();
+    List<BytesStreamMessage> messages;
     KinesisPartitionGroupOffset offsetOfNextBatch;
-    if (messages.isEmpty()) {
-      offsetOfNextBatch = startOffset;
-    } else {
+    if (!records.isEmpty()) {
+      messages = records.stream().map(record -> extractStreamMessage(record, shardId)).collect(Collectors.toList());
       StreamMessageMetadata lastMessageMetadata = messages.get(messages.size() - 1).getMetadata();
       assert lastMessageMetadata != null;
       offsetOfNextBatch = (KinesisPartitionGroupOffset) lastMessageMetadata.getNextOffset();
+    } else {
+      // TODO: Revisit whether Kinesis can return empty batch when there are available records. The consumer cna handle
+      //       empty message batch, but it will treat it as fully caught up.
+      messages = List.of();
+      offsetOfNextBatch = startOffset;
     }
-    return new KinesisMessageBatch(messages, offsetOfNextBatch, endOfShard);
+    assert offsetOfNextBatch != null;
+    _nextStartSequenceNumber = offsetOfNextBatch.getSequenceNumber();
+    _nextShardIterator = getRecordsResponse.nextShardIterator();
+    return new KinesisMessageBatch(messages, offsetOfNextBatch, _nextShardIterator == null);
   }
 
-  private String getShardIterator(String shardId, String sequenceNumber) {
-    GetShardIteratorRequest.Builder requestBuilder =
-        GetShardIteratorRequest.builder().streamName(_config.getStreamTopicName()).shardId(shardId);
-    if (sequenceNumber != null) {
-      requestBuilder = requestBuilder.startingSequenceNumber(sequenceNumber)
-          .shardIteratorType(ShardIteratorType.AFTER_SEQUENCE_NUMBER);
+  /**
+   * Kinesis enforces a limit of 5 getRecords request per second on each shard from AWS end, beyond which we start
+   * getting {@link ProvisionedThroughputExceededException}. Rate limit the requests to avoid this.
+   */
+  private void rateLimitRequests() {
```

Comment:
Thanks for creating a separate method. I guess this being a special kind of rate limiter that needs to block until we are ready to fetch again, we cannot leverage off the shelf ones like guava. 

if kinesis has a limit, don't we need to adhere to that limit. So does getRpsLimit() need to be what Kinesis limit is ?

Reply:
Kinesis [limit](https://docs.aws.amazon.com/streams/latest/dev/service-sizes-and-limits.html) is not very straight forward, so I guess we need to iterate on this to get the best settings.